[dtensor][view_op] add as_strided op support to DTensor in FakeTensorMode

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

* __->__ #151495

## Introduction
`flex_attention`'s FakeTensor propagation `flex_attention_fake_impl` [permutes](https://github.com/pytorch/pytorch/blob/fb6ac2f16132f7953711ce6924bc2ee4a033228c/torch/_higher_order_ops/flex_attention.py#L459) the stride of `out` (the attention score) based on `query`'s stride. To enable `flex_attention` call on DTensor, this requires us add `as_strided` support on DTensor in `FakeTensorMode`. 

## Limited Support
Due to the complexity of supporting actual `as_strided` on DTensor, I choose to only enable a limited subset:
1. `as_strided` only works correctly in `FakeTensorMode` i.e. shape and strided propagation.
2. `as_strided` is only allowed in case where `size == input.shape` because this PR specifically unblocks the use case of `flex_attention_fake_impl`. 
3. `as_strided` requires `storage_offset=None` because the other case is not defined in DTensor.

## Test
`pytest test/distributed/tensor/test_view_ops.py -s -k test_as_strided`

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @tianyu-l

XilunWu committed:
[dtensor][view_op] add as_strided op support to DTensor

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][view_op] add as_strided op support to DTensor in FakeTensorMode"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

*  #151497
*  #151496
* __->__ #151495

## Introduction
`flex_attention`'s FakeTensor propagation `flex_attention_fake_impl` [permutes](https://github.com/pytorch/pytorch/blob/fb6ac2f16132f7953711ce6924bc2ee4a033228c/torch/_higher_order_ops/flex_attention.py#L459) the stride of `out` (the attention score) based on `query`'s stride. To enable `flex_attention` call on DTensor, this requires us add `as_strided` support on DTensor in `FakeTensorMode`. 

## Limited Support
Due to the complexity of supporting actual `as_strided` on DTensor, I choose to only enable a limited subset:
1. `as_strided` only works correctly in `FakeTensorMode` i.e. shape and strided propagation.
2. `as_strided` is only allowed in case where `size == input.shape` because this PR specifically unblocks the use case of `flex_attention_fake_impl`. 
3. `as_strided` requires `storage_offset=None` because the other case is not defined in DTensor.

## Test
`pytest test/distributed/tensor/test_view_ops.py -s -k test_as_strided`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][view_op] add as_strided op support to DTensor in FakeTensorMode"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

*  #151497
*  #151496
* __->__ #151495

## Introduction
`flex_attention`'s FakeTensor propagation `flex_attention_fake_impl` [permutes](https://github.com/pytorch/pytorch/blob/fb6ac2f16132f7953711ce6924bc2ee4a033228c/torch/_higher_order_ops/flex_attention.py#L459) the stride of `out` (the attention score) based on `query`'s stride. To enable `flex_attention` call on DTensor, this requires us add `as_strided` support on DTensor in `FakeTensorMode`. 

## Limited Support
Due to the complexity of supporting actual `as_strided` on DTensor, I choose to only enable a limited subset:
1. `as_strided` only works correctly in `FakeTensorMode` i.e. shape and strided propagation.
2. `as_strided` is only allowed in case where `size == input.shape` because this PR specifically unblocks the use case of `flex_attention_fake_impl`. 
3. `as_strided` requires `storage_offset=None` because the other case is not defined in DTensor.

## Test
`pytest test/distributed/tensor/test_view_ops.py -s -k test_as_strided`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k tianyu-l

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_ops/_view_ops.py:
```diff
@@ -654,3 +654,45 @@ def reshape_strategy(op_schema: OpSchema) -> StrategyType:
+
+
+@register_op_strategy(aten.as_strided.default, schema_info=RuntimeSchemaInfo(1))
+def as_strided_strategy(op_schema: OpSchema) -> StrategyType:
```
Where do we enforce that this is only used in fake mode?

XilunWu replied:
this is the part I don't feel confident -- I did not enforce that. I only tested that it works out for my use case in Fake Tensor mode and totally rely on users to read the comment and not use it in other case.

wconstab reviewed (commented):
Can you explain further why we need this support only in fake mode? Is it because we trace through this op and the resulting graph needs to capture a local as_strided operation, but the operation would run on a plain tensor after compilation?

XilunWu committed:
Update on "[dtensor][view_op] add as_strided op support to DTensor in FakeTensorMode"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

*  #151497
*  #151507
* __->__ #151495

## Introduction
`flex_attention`'s FakeTensor propagation `flex_attention_fake_impl` [permutes](https://github.com/pytorch/pytorch/blob/fb6ac2f16132f7953711ce6924bc2ee4a033228c/torch/_higher_order_ops/flex_attention.py#L459) the stride of `out` (the attention score) based on `query`'s stride. To enable `flex_attention` call on DTensor, this requires us add `as_strided` support on DTensor in `FakeTensorMode`. 

## Limited Support
Due to the complexity of supporting actual `as_strided` on DTensor, I choose to only enable a limited subset:
1. `as_strided` only works correctly in `FakeTensorMode` i.e. shape and strided propagation.
2. `as_strided` is only allowed in case where `size == input.shape` because this PR specifically unblocks the use case of `flex_attention_fake_impl`. 
3. `as_strided` requires `storage_offset=None` because the other case is not defined in DTensor.

## Test
`pytest test/distributed/tensor/test_view_ops.py -s -k test_as_strided`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k tianyu-l

[ghstack-poisoned]

XilunWu committed:
Update on "[dtensor][view_op] add as_strided op support to DTensor in FakeTensorMode"

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):

*  #151497
*  #151507
* __->__ #151495

## Introduction
`flex_attention`'s FakeTensor propagation `flex_attention_fake_impl` [permutes](https://github.com/pytorch/pytorch/blob/fb6ac2f16132f7953711ce6924bc2ee4a033228c/torch/_higher_order_ops/flex_attention.py#L459) the stride of `out` (the attention score) based on `query`'s stride. To enable `flex_attention` call on DTensor, this requires us add `as_strided` support on DTensor in `FakeTensorMode`. 

## Limited Support
Due to the complexity of supporting actual `as_strided` on DTensor, I choose to only enable a limited subset:
1. `as_strided` only works correctly in `FakeTensorMode` i.e. shape and strided propagation.
2. `as_strided` is only allowed in case where `size == input.shape` because this PR specifically unblocks the use case of `flex_attention_fake_impl`. 
3. `as_strided` requires `storage_offset=None` because the other case is not defined in DTensor.

## Test
`pytest test/distributed/tensor/test_view_ops.py -s -k test_as_strided`

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k tianyu-l

[ghstack-poisoned]

wanchaol reviewed (changes requested):
see comments inlined
